File: mobile/lib/models/product.dart
class Product {
  Product({
    required this.id,
    required this.name,
    required this.priceCents,
    this.sku,
    this.barcode,
    this.costCents = 0,
    this.stockQty = 0,
    this.reorderLevel = 0,
    this.unit = 'pc',
    this.vatClass = 'STANDARD',
    this.deleted = false,
    this.version = 1,
    DateTime? updatedAt,
    this.dirty = false,
  }) : updatedAt = updatedAt ?? DateTime.now().toUtc();

  final String id;
  String? sku;
  String name;
  String? barcode;
  int priceCents;
  int costCents;
  num stockQty;
  num reorderLevel;
  String unit;
  String vatClass;
  bool deleted;
  int version;
  DateTime updatedAt;
  bool dirty;

  double get price => priceCents / 100.0;

  Map<String, dynamic> toMap() => {
        'id': id,
        'sku': sku,
        'name': name,
        'barcode': barcode,
        'priceCents': priceCents,
        'costCents': costCents,
        'stockQty': stockQty,
        'reorderLevel': reorderLevel,
        'unit': unit,
        'vatClass': vatClass,
        'deleted': deleted,
        'version': version,
        'updatedAt': updatedAt.toIso8601String(),
        'dirty': dirty,
      };

  Map<String, dynamic> toApiPayload() => {
        'id': id,
        'sku': sku,
        'name': name,
        'barcode': barcode,
        'priceCents': priceCents,
        'costCents': costCents,
        'stockQty': stockQty,
        'reorderLevel': reorderLevel,
        'unit': unit,
        'vatClass': vatClass,
        'deleted': deleted,
      };

  factory Product.fromMap(Map<String, dynamic> m) => Product(
        id: m['id'] as String,
        sku: m['sku'] as String?,
        name: m['name'] as String,
        barcode: m['barcode'] as String?,
        priceCents: (m['priceCents'] as num).toInt(),
        costCents: (m['costCents'] as num?)?.toInt() ?? 0,
        stockQty: (m['stockQty'] as num?) ?? 0,
        reorderLevel: (m['reorderLevel'] as num?) ?? 0,
        unit: m['unit'] as String? ?? 'pc',
        vatClass: m['vatClass'] as String? ?? 'STANDARD',
        deleted: m['deleted'] as bool? ?? false,
        version: (m['version'] as num?)?.toInt() ?? 1,
        updatedAt: DateTime.tryParse(m['updatedAt'] as String? ?? '')?.toUtc() ??
            DateTime.now().toUtc(),
        dirty: m['dirty'] as bool? ?? false,
      );
}

File: backend/src/main/java/com/tillpro/api/domain/Product.java
package com.tillpro.api.domain;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.UUID;

@Entity
@Table(name = "products",
       uniqueConstraints = @UniqueConstraint(columnNames = {"tenant_id", "sku"}))
@Getter @Setter @NoArgsConstructor @AllArgsConstructor @Builder
public class Product {
    @Id
    @Column(columnDefinition = "uuid")
    private UUID id;

    @Column(name = "tenant_id", nullable = false)
    private UUID tenantId;

    private String sku;

    @Column(nullable = false)
    private String name;

    private String barcode;

    @Column(name = "price_cents", nullable = false)
    private Long priceCents;

    @Column(name = "cost_cents", nullable = false)
    private Long costCents = 0L;

    @Column(name = "stock_qty", nullable = false, precision = 14, scale = 3)
    private BigDecimal stockQty = BigDecimal.ZERO;

    @Column(name = "reorder_level", nullable = false, precision = 14, scale = 3)
    private BigDecimal reorderLevel = BigDecimal.ZERO;

    @Column(nullable = false)
    private String unit = "pc";

    @Column(name = "vat_class", nullable = false)
    private String vatClass = "STANDARD";

    @Column(nullable = false)
    private Boolean deleted = false;

    @Version
    private Long version;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private Instant createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private Instant updatedAt;
}
